includes/kaltura_derivatives.inc

```diff
@@ -26,28 +50,44 @@ function islandora_video_streaming_kaltura_upload($object, $force = false) {
     try {
       $helpers = new KalturaHelpers();
       $client = $helpers->getKalturaClient();
+      
+      if (isset($old_content))
+      {
+        try {
+          $myService = new KalturaBaseEntryService($client);
+          $myService->delete($old_entity_id);
+        } catch (Exception $ex) {
+          $failureMessage['messages'][0]['message'] = t('KalturaBaseEntry error: @exception');
+          $failureMessage['messages'][0]['message_sub']['@exception'] = $ex->getMessage();
+          watchdog_exception('islandora_video_streaming_kaltura', $ex);
+        }
+      }
 
-      // TODO: would it be better or worse to skip the KalturaMediaEntry ?
       $entry = new KalturaMediaEntry();
       $entry->name = pathinfo($filePath, PATHINFO_FILENAME);
-      $entry->mediaType = kaltura_get_media_type_by_mime(file_get_mimetype($filePath));
+      $entry->mediaType = kaltura_get_media_type_by_mime(file_get_mimetype($filePath)); //original datafile
 
+      //returns token of video and uploads to kaltura
       $token = $client->media->upload(drupal_realpath($filePath));
       $entry = $client->media->addFromUploadedFile($entry, $token);
 
       $entity = kaltura_entry_create(array('kaltura_entryid' => $entry->id));
       kaltura_map_properties($entry, $entity);
-      // TODO: duplicate metadata via kaltura_save_entry_metadata() ?
-      kaltura_entry_save($entity);
+            
+      kaltura_save_entry_metadata($entity);
```

Comment:
Islandora supplies us with rich metadata, likely in the MODS or DC datastream.  I don't know if we are guaranteed a MODS or DC at the time of processing this upload, though.  What kinds of metadata (title, format, length, etc.?) does `kaltura_entry_save` expect to be recorded in the `$entity`?  If this is accessible, we should try to add it.

Reply:
_What kinds of metadata (title, format, length, etc.?) does kaltura_entry_save expect to be recorded in the $entity?_  Unknown

Reply:
_What kinds of metadata (title, format, length, etc.?) does kaltura_entry_save expect to be recorded in the $entity?_ Unknown, the function kaltura_entry_save uses function entity_save.

Reply:
Sorry, should have referenced `kaltura_save_entry_metadata()` here, not `kaltura_entry_save'.  Take a look at how metadata is leveraged in the Kaltura Media Entries in Drupal.  If there is metadata about the conceptual object (title, description) or about the digital format (length, format), we can potentially add it here.  Otherwise, we may not even need to call `kaltura_save_entry_metadata()`.

Reply:
The metadata about the Kaltura Media Entries only contain title, descriptions, and tags.

Reply:
Every Islandora Object has a formal title and description, as well as many subject and keyword elements which might be considered Kaltura "tags".  I think our goal with the Kaltura metadata should be twofold: 1) to apply a human readable element to the video, and 2) to backlink into islandora.  For that purpose, let's copy the Islandora Object's "Label" into the Kaltura Description, and the Islandora Object's PID (Persistent Identifier) into the Kaltura Title.  This should, in a Kaltura interface, present the Videos by Islandora Object, but also provide a human-readable Description.